src/IECore/MeshAlgo.cpp

```diff
@@ -1,16 +1,238 @@
+
+	MeshFaceVaryingToVertex( const MeshPrimitive *mesh )	:	m_mesh( mesh )
+	{
+	}
+
+	template<typename From> ReturnType operator()( const From* data )
+	{
+		typename From::Ptr result = static_cast< From* >( Object::create( data->typeId() ).get() );
+		typename From::ValueType &trg = result->writable();
+		const typename From::ValueType &src = data->readable();
+
+		size_t numVerts = m_mesh->variableSize( PrimitiveVariable::Vertex );
+		std::vector<int> count( numVerts, 0 );
+		trg.resize( numVerts );
+
+		const std::vector<int>& vertexIds = m_mesh->vertexIds()->readable();
+		std::vector<int>::const_iterator vertexIdIt = vertexIds.begin();
+		typename From::ValueType::const_iterator srcIt = src.begin(), srcEnd = src.end();
+
+		for( ; srcIt != srcEnd ; ++srcIt, ++vertexIdIt )
+		{
+			trg[ *vertexIdIt ] += *srcIt;
+			++count[ *vertexIdIt ];
+		}
+
+		std::vector<int>::const_iterator cIt = count.begin();
+		typename From::ValueType::iterator trgIt = trg.begin(), trgEnd = trg.end();
+		for( trgIt = trg.begin(); trgIt != trgEnd ; ++trgIt, ++cIt )
+		{
+			*trgIt /= *cIt;
+		}
+
+		return result;
+	}
+
+	const MeshPrimitive *m_mesh;
+};
+
+struct MeshFaceVaryingToUniform
+{
+	typedef DataPtr ReturnType;
+
+	MeshFaceVaryingToUniform( const MeshPrimitive *mesh )	:	m_mesh( mesh )
+	{
+	}
+
+	template<typename From> ReturnType operator()( const From* data )
+	{
+		typename From::Ptr result = static_cast< From* >( Object::create( data->typeId() ).get() );
+		typename From::ValueType &trg = result->writable();
+		const typename From::ValueType &src = data->readable();
+
+		trg.reserve( m_mesh->numFaces() );
+
+		const std::vector<int>& vertexIds = m_mesh->vertexIds()->readable();
+		typename From::ValueType::const_iterator srcIt = src.begin();
+
+		const std::vector<int> &verticesPerFace = m_mesh->verticesPerFace()->readable();
+		for( std::vector<int>::const_iterator it = verticesPerFace.begin(); it < verticesPerFace.end(); ++it )
+		{
+			// initialize with the first value to avoid
+			// ambiguity during default construction
+			typename From::ValueType::value_type total = *srcIt;
+			++srcIt;
+
+			for( int j = 1; j < *it; ++j, ++srcIt )
+			{
+				total += *srcIt;
+			}
+
+			trg.push_back( total / *it );
+		}
+
+		return result;
+	}
+
+	const MeshPrimitive *m_mesh;
+};
+
+struct MeshAnythingToFaceVarying
+{
+	typedef DataPtr ReturnType;
+
+	MeshAnythingToFaceVarying( const MeshPrimitive *mesh, PrimitiveVariable::Interpolation srcInterpolation )
+		: m_mesh( mesh ), m_srcInterpolation( srcInterpolation )
+	{
+	}
+
+	template<typename From> ReturnType operator()( const From* data )
+	{
+		// we need to duplicate because the Op expects a primvar to manipulate..
+		IECore::MeshPrimitivePtr tmpMesh = m_mesh->copy();
+		// cast OK due to read-only access.
+		tmpMesh->variables["tmpPrimVar"] = IECore::PrimitiveVariable( m_srcInterpolation, const_cast< From * >(data) );
+		IECore::FaceVaryingPromotionOpPtr promoteOp = new IECore::FaceVaryingPromotionOp();
```

Comment:
I presume `MeshAlgo::resamplePrimitiveVariable()` will supersede FaceVaryingPromotionOp totally in the next major version, at which point we'll just move those internals here and delete the op?